src/storage/v2/inmemory/storage.hpp

```diff
@@ -0,0 +1,469 @@
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#pragma once
+
+#include "storage/v2/replication/config.hpp"
+#include "storage/v2/replication/enums.hpp"
+#include "storage/v2/storage.hpp"
+
+namespace memgraph::storage {
+
+// The storage is based on this paper:
+// https://db.in.tum.de/~muehlbau/papers/mvcc.pdf
+// The paper implements a fully serializable storage, in our implementation we
+// only implement snapshot isolation for transactions.
+
+class InMemoryStorage final : public Storage {
+ public:
+  enum class RegisterReplicaError : uint8_t {
+    NAME_EXISTS,
+    END_POINT_EXISTS,
+    CONNECTION_FAILED,
+    COULD_NOT_BE_PERSISTED
+  };
+
+  struct TimestampInfo {
+    uint64_t current_timestamp_of_replica;
+    uint64_t current_number_of_timestamp_behind_master;
+  };
+
+  enum class ReplicationRole : uint8_t { MAIN, REPLICA };
+
+  struct ReplicaInfo {
+    std::string name;
+    replication::ReplicationMode mode;
+    io::network::Endpoint endpoint;
+    replication::ReplicaState state;
+    TimestampInfo timestamp_info;
+  };
+
+  enum class CreateSnapshotError : uint8_t {
+    DisabledForReplica,
+    DisabledForAnalyticsPeriodicCommit,
+    ReachedMaxNumTries
+  };
+
+  /// @throw std::system_error
+  /// @throw std::bad_alloc
+  explicit InMemoryStorage(Config config = Config());
+
+  InMemoryStorage(const InMemoryStorage &) = delete;
+  InMemoryStorage(InMemoryStorage &&) = delete;
+  InMemoryStorage &operator=(const InMemoryStorage &) = delete;
+  InMemoryStorage &operator=(InMemoryStorage &&) = delete;
+
+  ~InMemoryStorage() override;
+
+  class InMemoryAccessor final : public Storage::Accessor {
+   private:
+    friend class InMemoryStorage;
+
+    explicit InMemoryAccessor(InMemoryStorage *storage, IsolationLevel isolation_level, StorageMode storage_mode);
+
+   public:
+    InMemoryAccessor(const InMemoryAccessor &) = delete;
+    InMemoryAccessor &operator=(const InMemoryAccessor &) = delete;
+    InMemoryAccessor &operator=(InMemoryAccessor &&other) = delete;
+
+    // NOTE: After the accessor is moved, all objects derived from it (accessors
+    // and iterators) are *invalid*. You have to get all derived objects again.
+    InMemoryAccessor(InMemoryAccessor &&other) noexcept;
+
+    ~InMemoryAccessor() override;
+
+    /// @throw std::bad_alloc
+    VertexAccessor CreateVertex() override;
+
+    std::optional<VertexAccessor> FindVertex(Gid gid, View view) override;
+
+    VerticesIterable Vertices(View view) override {
+      auto *mem_storage = static_cast<InMemoryStorage *>(storage_);
```

Comment:
why do we need static_cast here? Should proper call be resolved by function call?